### PR TITLE
Add simple vote scores

### PR DESCRIPTION
**Pull request by matthewkol186**
(See issue #25.) Also, **N.B.**: I've added a lot of explanatory comments, because I suspect this PR may be hard to parse!

Currently, the `ElectionTracker` object has a `DistrictScore` baked in that recalculates vote counts & vote shares for every party. This `DistrictScore` both updates the `Election` object as well as returns a `Dict` that maps party name to a tuple of vote count & vote share won by that party in a particular district. This is non-ideal, and also makes it difficult for users to easily extract the values of vote count/vote share by party from the object returned by `recom_chain()` / `flip_chain`. Furthermore, it gave no way for the user to only track vote count/share for a particular party of interest, instead forcing the user to record outcomes for all parties.

This PR splits the existing `count_votes` function (which creates a `DistrictScore` that updates the `Election` object and returns that big `Dict` of vote counts/shares) into three functions: `vote_updater` (which only updates the `Election` object and is not visible to the user), `vote_count` (which returns a `DistrictScore` that calculates the vote count for a particular party in a particular district), and `vote_share` (which returns a `DistrictScore` that calculates the vote share for a particular party in a particular district).

### Usage, before change
```
election = Election("SEN10", ["SEN10D", "SEN10R"], graph.num_dists)
election_metrics = [   # optional
    efficiency_gap("efficiency_gap", election, "SEN10D"),
    seats_won("seats_won", election, "SEN10D"),
]
...
scores = [
    ...
    ElectionTracker(election, partisan_metrics=election_metrics)
    ...
]
...
score_values = recom_chain(graph, partition, population_constraint, num_steps, scores)
# Uh oh, the votes for D are somewhere in the return object, but there's no method for me to easily retrieve it...
```

### Usage, after change
```
election = Election("SEN10", ["SEN10D", "SEN10R"], graph.num_dists)
election_metrics = [   # optional
    vote_share("share_d", election, "SEN10D"), # now user can specify what party they want to explicitly track results for!
    efficiency_gap("efficiency_gap", election, "SEN10D"),
    seats_won("seats_won", election, "SEN10D"),
]
...
scores = [
    ...
    ElectionTracker(election, election_metrics)
    ...
]
...
score_values = recom_chain(graph, partition, population_constraint, num_steps, scores)
d_share = get_score_values(score_values, "share_d") # yay! now it's easier to get this data
```

**PROMISE**: Once this is approved, I promise to edit the documentation appropriately and only merge once I am finished!

### CODE REVIEW

File: src/election.jl [hunk 1]
```diff
@@ -15,12 +15,12 @@ function Election(name::String, parties::Array{String, 1}, num_districts::Int)
     return Election(name, parties, vote_counts, vote_shares)
 end
 
-function count_votes(name::String, election::Election)::DistrictScore
```
Comment (matthewkol186) [inline src/election.jl:18]: This function was doing triple duty! Updating the `Election` object and returning the vote counts and vote shares for all parties in the district.

File: src/election.jl [hunk 2]
```diff
@@ -41,14 +41,37 @@ function count_votes(name::String, election::Election)::DistrictScore
         vote_totals = sum(election.vote_counts[district, :])
         vote_shares = election.vote_counts[district, :] ./ vote_totals
         election.vote_shares[district, :] = vote_shares
```
Comment (matthewkol186) [inline src/election.jl:43]: Updating the `Election` object (now in the new code, this is only performed by `vote_updater`)

File: src/election.jl [hunk 3]
```diff
@@ -41,14 +41,37 @@ function count_votes(name::String, election::Election)::DistrictScore
...
-        # create Dict matching party to vote count & share
-        vote_pairs = Dict{String, NamedTuple}()
-        for i in 1:length(party_names)
-            party_votes = election.vote_counts[district, i]
-            party_share = election.vote_shares[district, i]
-            vote_pairs[party_names[i]] = (vote_count=party_votes, vote_share=party_share)
-        end
-        return vote_pairs
```
Comment (matthewkol186) [inline src/election.jl:51]: Returning a `Dict` that stores all vote count/vote shares for every party in this particular district. Now, in the new code, this is done separately by `vote_count` and `vote_share`, and only for the parties that the user requests!

File: src/election.jl [hunk 4]
```diff
@@ -172,17 +195,18 @@ end
...
-        partisan metrics. Furthermore, packaging all partisan metrics within
-        the CompositeScore ensures that the vote update occurs first, followed
-        by the partisan metrics scoring functions.
+        partisan metrics. Furthermore, packaging all election-related scores
+        within the CompositeScore ensures that the vote update occurs first,
+        followed by the partisan metrics scoring functions.
     """
-    vote_tracker = [count_votes("votes", election)]
-    scores = Array{AbstractScore, 1}([vote_tracker; partisan_metrics])
+    count_votes = vote_updater("votes", election) # name does not matter
```
Comment (matthewkol186) [inline src/election.jl:209]: Note: name doesn't matter because I've also updated `score_initial_partition` and `score_partition_from_proposal` to only store the results of score functions that return an actual value (i.e., they do not return `nothing`). Since the `vote_updater` function's purpose is only to update the `Election` object (not to actually return a value of a score), it will not appear in the object returned by `flip_chain()`/`recom_chain()`, and therefore the name we provide here has no real significance.

File: src/scores.jl [hunk 5]
```diff
@@ -217,15 +218,22 @@ function score_partition_from_proposal(graph::BaseGraph,
...
-            score_values[s.name] = score_partition_from_proposal(graph, partition, proposal, s.scores)
-            delete!(score_values[s.name], "dists") # remove redundant dists key
+            value = score_partition_from_proposal(graph, partition, proposal, s.scores)
+            delete!(value, "dists") # remove redundant dists key
         else # efficiently calculate & store scores only on changed districts
-            score_values[s.name] = eval_score_on_districts(graph, partition, s, Δ_districts)
+            value = eval_score_on_districts(graph, partition, s, Δ_districts)
+            empty_return = value[1] == nothing || value[2] == nothing
+        end
+        if !empty_return
```
Comment (matthewkol186) [inline src/scores.jl:235]: Ensures that if the value returned by the score was empty (i.e., `nothing` or an Array of `nothing`), it does not get written to the score value dictionary.

File: src/election.jl [hunk 6]
```diff
@@ -172,17 +195,18 @@ end
...
         Re-calculating vote counts only for changed districts means that the
         CompositeScore does not perform redundant computations for all of the
-        partisan metrics. Furthermore, packaging all partisan metrics within
-        the CompositeScore ensures that the vote update occurs first, followed
-        by the partisan metrics scoring functions.
+        partisan metrics. Furthermore, packaging all election-related scores
+        within the CompositeScore ensures that the vote update occurs first,
+        followed by the partisan metrics scoring functions.
     """
-    vote_tracker = [count_votes("votes", election)]
```
Comment (matthewkol186) [inline src/election.jl:185]: Previously, if you wanted to figure out the votes for Party A throughout the chain, you had to do `get_score_values(chain_data, "votes")` and then process an array of `Dict` objects...which makes no sense to the user because they never initialie a score with the name `"votes"` - that's something we do internally to this method!

File: src/scores.jl [hunk 7]
```diff
@@ -217,15 +218,22 @@ function score_partition_from_proposal(graph::BaseGraph,
...
             # ensure that district-level scores in the CompositeScore are only
             # evaluated on changed districts
-            score_values[s.name] = score_partition_from_proposal(graph, partition, proposal, s.scores)
-            delete!(score_values[s.name], "dists") # remove redundant dists key
+            value = score_partition_from_proposal(graph, partition, proposal, s.scores)
+            delete!(value, "dists") # remove redundant dists key
         else # efficiently calculate & store scores only on changed districts
-            score_values[s.name] = eval_score_on_districts(graph, partition, s, Δ_districts)
+            value = eval_score_on_districts(graph, partition, s, Δ_districts)
+            empty_return = value[1] == nothing || value[2] == nothing
```
Comment (bsuwal) [inline src/scores.jl:233]: is this the check to ensure everything in the array is `nothing`? It's only checking the first two values

**Comment by bsuwal** [comment]
also wanted to say thanks for the comments on the PR @matthewkol186 , they were very helpful!

File: src/scores.jl [hunk 8]
```diff
@@ -187,7 +183,12 @@ function score_initial_partition(graph::BaseGraph,
     """
     score_values = Dict{String, Any}()
     for s in scores
-        score_values[s.name] = eval_score_on_partition(graph, partition, s)
+         value = eval_score_on_partition(graph, partition, s)
+         # check if value returned by score was empty
+         val_empty = value == nothing
+         # check if array of values returned by score was empty
+         val_empty |= (value isa Array && count(x -> x == nothing, value) == length(value))
```
[the remark this answers is not in the document]
Comment (matthewkol186) [reply src/scores.jl:190]: Good q! An `Array` of `nothing` is returned when you run `eval_score_on_partition` with a district-level score. This returns an `Array` of values, where each value represents the return value of the district-level score on a particular district. For example, the `Election`'s `vote_updater` is a district-level score that returns `nothing`, so when you run `eval_score_on_partition(graph, partition, an `Array` of `nothing` gets returned.

[the remark this answers is not in the document]
Comment (matthewkol186) [reply src/scores.jl:190]: In the future, I'd like to figure out a better way to do this check more systematically, although I think this is fine for now. See #30!

File: src/scores.jl (hunk 7 above)
Comment (matthewkol186) [reply src/scores.jl:233]: Because we know `Δ_districts` is literally defined to be `[proposal.D₁, proposal.D₂]`, `eval_score_on_districts(graph, partition, s, Δ_districts)` returns an array of length 2. A lot of our codebase is written explicitly assuming that `dists` is length 2. That being said, I changed the check to use `any` for posterity, since in the future `dists` is likely to change to accommodate >2 changed districts!

File: src/scores.jl (hunk 8 above)
[the remark this answers is not in the document]
Comment (bsuwal) [reply src/scores.jl:190]: sounds good!

File: src/scores.jl (hunk 7 above)
Comment (bsuwal) [reply src/scores.jl]: sounds good! hadn't realized it was on `\del_districts` thank you!
